packages/frontend-app-api/src/wiring/createApp.test.tsx

```diff
@@ -0,0 +1,299 @@
+ */
+
+import {
+  createExtension,
+  createExtensionDataRef,
+  createExtensionInput,
+  createPageExtension,
+  createPlugin,
+} from '@backstage/frontend-plugin-api';
+import { createInstances } from './createApp';
+
+import { MockConfigApi } from '@backstage/test-utils';
+import React from 'react';
+import { createRouteRef } from '@backstage/core-plugin-api';
+
+describe('createInstances', () => {
+  it('throws an error when a root extension is parametrized', () => {
+    const config = new MockConfigApi({
+      app: {
+        extensions: [
+          {
+            root: {
+              at: '',
+            },
+          },
+        ],
+      },
+    });
+    const plugins = [
+      createPlugin({
+        id: 'plugin',
+        extensions: [],
+      }),
+    ];
+    expect(() => createInstances({ config, plugins })).toThrow(
+      "A 'root' extension configuration was detected, but the root extension is not configurable",
+    );
+  });
+
+  it('throws an error when a root extension is overridden', () => {
+    const config = new MockConfigApi({});
+    const plugins = [
+      createPlugin({
+        id: 'plugin',
+        extensions: [
+          createExtension({
+            id: 'root',
+            at: 'core.routes/route',
+            inputs: {},
+            output: {},
+            factory() {},
+          }),
+        ],
+      }),
+    ];
+    expect(() => createInstances({ config, plugins })).toThrow(
+      "The following plugin(s) are overriding the 'root' extension which is forbidden: plugin",
+    );
+  });
+
+  describe('should deduplicated plugins', () => {
+    it('in an immediate dependency level (e.g., A -> A)', () => {
+      const config = new MockConfigApi({});
+
+      const addonExtensionData =
+        createExtensionDataRef<JSX.Element>('plugin.page.addon');
+
+      const CyclicalA = createExtension({
+        id: 'A', // Same id as ancestor A
+        at: 'A/addons',
+        inputs: {},
+        output: {
+          element: addonExtensionData,
+        },
+        factory({ bind }) {
+          bind({
+            element: <div key="cyclical-a">Cyclical A</div>,
+          });
+        },
+      });
+
+      const A = createPageExtension({
+        id: 'A',
+        defaultPath: '/',
+        routeRef: createRouteRef({ id: 'A.route' }),
+        inputs: {
+          addons: createExtensionInput({
+            element: addonExtensionData,
+          }),
+        },
+        loader: async ({ inputs }) => (
+          <div>A {inputs.addons.map(({ element }) => element)}</div>
+        ),
+      });
+
+      const plugins = [
+        createPlugin({
+          id: 'plugin',
+          extensions: [A, CyclicalA],
+        }),
+      ];
+
+      // It should not create an infinite loop
+      expect(() => createInstances({ config, plugins })).not.toThrow(
+        'Maximum call stack size exceeded',
+      );
+    });
+
+    it('in an intermediate level (e.g., A -> B -> A)', () => {
```

Comment:
We can remove this and the next test now right? It's basically the same as the above one. Always nice to have less code 😁

Reply:
Removed 🙌🏻 : https://github.com/backstage/backstage/pull/20032/commits/c71da2a7907c53e6d147b76925ee0f1341406669